source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -272,111 +272,183 @@ absl::Status ProcessorState::handleHeadersResponse(const HeadersResponse& respon
+  // Handle body mutations if present
+  if (common_response.has_body_mutation()) {
+    const auto validation_status = validateContentLength(common_response);
+    if (!validation_status.ok()) {
+      return validation_status;
     }
+    applyBufferedBodyMutation(common_response);
+  }
+
+  clearWatermark();
+  onFinishProcessorCall(Grpc::Status::Ok);
+  should_continue = true;
+  return absl::OkStatus();
+}
+
+absl::Status ProcessorState::handleStreamedBodyCallback(const CommonResponse& common_response,
+                                                        bool& should_continue) {
+  absl::StatusOr<bool> result = handleBodyInStreamedState(common_response);
+  if (!result.ok()) {
+    return result.status();
+  }
+  should_continue = *result;
+  return absl::OkStatus();
+}
+
+absl::Status
+ProcessorState::handleBufferedPartialBodyCallback(const CommonResponse& common_response,
+                                                  bool& should_continue) {
+  Buffer::OwnedImpl chunk_data;
+  auto chunk = dequeueStreamingChunk(chunk_data);
+  if (!chunk) {
+    ENVOY_BUG(false, "Bad partial body callback state");
+    return absl::InternalError("Invalid chunk in partial body callback");
+  }
 
-    if (should_continue || (trailers_available_ && chunk_queue_.empty())) {
-      continueIfNecessary();
+  // Process header mutations if present
+  if (common_response.has_header_mutation()) {
+    const auto mut_status = processHeaderMutationIfAvailable(common_response);
+    if (!mut_status.ok()) {
+      return mut_status;
     }
+  }
+
+  // Apply body mutations and process data
+  if (common_response.has_body_mutation()) {
+    MutationUtils::applyBodyMutations(common_response.body_mutation(), chunk_data);
+  }
+
+  processChunkDataAndLeftovers(chunk_data, chunk->end_stream);
+
+  should_continue = true;
+  clearWatermark();
+  onFinishProcessorCall(Grpc::Status::Ok);
+  partial_body_processed_ = true;
+
+  return absl::OkStatus();
+}
+
+absl::Status
+ProcessorState::processHeaderMutationIfAvailable(const CommonResponse& common_response) {
+  if (headers_ != nullptr) {
+    return processHeaderMutation(common_response);
+  }
+  ENVOY_LOG(debug, "Response had header mutations but headers aren't available");
+  return absl::OkStatus();
+}
+
+absl::Status ProcessorState::validateContentLength(const CommonResponse& common_response) {
+  if (!headers_ || !headers_->ContentLength()) {
     return absl::OkStatus();
   }
 
-  return absl::FailedPreconditionError("spurious message");
+  size_t content_length = 0;
+  if (!absl::SimpleAtoi(headers_->getContentLengthValue(), &content_length)) {
+    return absl::OkStatus();
+  }
+
+  if (content_length != common_response.body_mutation().body().size()) {
+    return absl::InternalError(
+        "mismatch between content length and the length of the mutated body");
+  }
+
+  return absl::OkStatus();
+}
+
+void ProcessorState::applyBufferedBodyMutation(const CommonResponse& common_response) {
```

Comment:
This function is called only in one place.  Why not just put these a few lines inline its caller?

Reply:
IMO it's OK to have a single-use function, even if it's not huge, if it makes call-sites more readable. I'm not sure if that's the case here.